chainerx_cc/chainerx/routines/math.cc

```diff
@@ -594,28 +593,47 @@ Array ElementwisePower(const Array& x1, const Array& x2) {
     {
         BackwardBuilder bb{"power", {x1, broadcasted_x2}, out};
         if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
-            bt.Define([x1_tok = bb.RetainInput(0), x2_tok = bb.RetainInput(1)](BackwardContext& bctx) {
+            bt.Define([x1_tok = bb.RetainInput(0), x2_tok = bb.RetainInput(1), out_tok = bb.RetainOutput(0)](BackwardContext& bctx) {
+                const Array& gout = *bctx.output_grad();
                 const Array& x1 = bctx.GetRetainedInput(x1_tok);
                 const Array& x2 = bctx.GetRetainedInput(x2_tok);
-                Array pow_x1 = EmptyLike(x1, x1.device());
-                x1.device().Pow(x1, x2 - 1, pow_x1);
-
-                bctx.input_grad() = *bctx.output_grad() * x2 * pow_x1;
+                const Array& out = bctx.GetRetainedOutput(out_tok);
+                
+                bctx.input_grad() = gout * x2 * Pow(x1, x2 - 1);
             });
         }
         if (BackwardBuilder::Target bt = bb.CreateTarget(1)) {
             bt.Define([x1_tok = bb.RetainInput(0), out_tok = bb.RetainOutput(0)](BackwardContext& bctx) {
+                const Array& gout = *bctx.output_grad();
                 const Array& x1 = bctx.GetRetainedInput(x1_tok);
                 const Array& out = bctx.GetRetainedOutput(out_tok);
-                Array logx1 = EmptyLike(x1, x1.device());
-                x1.device().Log(x1, logx1);
 
-                bctx.input_grad() = *bctx.output_grad() * out * logx1;
+                bctx.input_grad() = gout * out * Log(x1); 
```

Comment:
I have done as you suggested, but still, the double backward test is failing.

Reply:
Would it be possible to provide some info about e.g. inputs and an actual failure message?

Reply:
The inputs are - 
a = [[2, 6],[4,6]] b = [[3, 5],[4, 2]]

Here is the test error - 
`
[ RUN      ] ForEachBackend/MathTest.PowerDoubleBackward/0
2: unknown file: Failure
2: C++ exception with description "Numerical error in backward on inputs (out of 3): 0, 1
2: Backprop ID: <default>
2: Atol: 1e-05  Rtol: 0.0001
2: Error[0]:
2: array([[-0.19791481, -0.        ],
2:        [28.06966120, -0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Backward gradients[0]:
2: array([[   24.30505552,    -0.        ],
2:        [-6255.52749001,     0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Numerical gradients[0]:
2: array([[   24.50297033,     0.        ],
2:        [-6283.59715121,     0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Eps[0] (perturbation in numerical gradients):
2: array([[0.1, 0.1],
2:        [0.1, 0.1]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Error[1]:
2: array([[ -0.05591017,  -0.        ],
2:        [124.21004340,  -0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Backward gradients[1]:
2: array([[   13.59493467,    -0.        ],
2:        [-8999.68248524,     0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Numerical gradients[1]:
2: array([[   13.65084483,     0.        ],
2:        [-9123.89252864,     0.        ]], shape=(2, 2), dtype=float64, device='cuda:0')
2: Eps[1] (perturbation in numerical gradients):
2: array([[0.1, 0.1],
2:        [0.1, 0.1]], shape=(2, 2), dtype=float64, device='cuda:0')
2: " thrown in the test body.
2: [  FAILED  ] ForEachBackend/MathTest.PowerDoubleBackward/0, where GetParam() = "cuda" (13 ms)
2: [ RUN      ] ForEachBackend/MathTest.PowerDoubleBackward/1
2: unknown file: Failure
2: C++ exception with description "Numerical error in backward on inputs (out of 3): 0, 1
2: Backprop ID: <default>
2: Atol: 1e-05  Rtol: 0.0001
2: Error[0]:
2: array([[-0.19791481, -0.        ],
2:        [28.06966120, -0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Backward gradients[0]:
2: array([[   24.30505552,    -0.        ],
2:        [-6255.52749001,     0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Numerical gradients[0]:
2: array([[   24.50297033,     0.        ],
2:        [-6283.59715121,     0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Eps[0] (perturbation in numerical gradients):
2: array([[0.1, 0.1],
2:        [0.1, 0.1]], shape=(2, 2), dtype=float64, device='native:0')
2: Error[1]:
2: array([[ -0.05591017,  -0.        ],
2:        [124.21004340,  -0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Backward gradients[1]:
2: array([[   13.59493467,    -0.        ],
2:        [-8999.68248524,     0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Numerical gradients[1]:
2: array([[   13.65084483,     0.        ],
2:        [-9123.89252864,     0.        ]], shape=(2, 2), dtype=float64, device='native:0')
2: Eps[1] (perturbation in numerical gradients):
2: array([[0.1, 0.1],
2:        [0.1, 0.1]], shape=(2, 2), dtype=float64, device='native:0')
2: " thrown in the test body.
2: [  FAILED  ] ForEachBackend/MathTest.PowerDoubleBackward/1, where GetParam() = "native" (4 ms)

`

Reply:
Fixed the issue.

Reply:
@dido1998 could you specify what the issue was with the tests earlier i was trying to get a workaround but couldn't.